### ISSUE TITLE
Synchronous and Asynchronous programming models

### ISSUE THREAD

**Issue by TheJebForge** [Yellow-Dog-Man/Resonite-Issues#829]
### Is your feature request related to a problem? Please describe.

As it currently stands, Resonite has sync and async protoflux nodes, it creates a lot of confusion for non-programmers and it's hard to work with in general.

As I understand, benefits of synchronous programming is that when you wait a frame, you will wait exactly a frame and your code will perform right on time since there's no task scheduler that needs to put your coroutine back into execution. But we don't even have any synchronous waits. 

Asynchronous code executes synchronously until it yields (a delay happens, something happens that waits for a thing to happen, etc), but sync nodes can be run inside of async tasks, so why do we still have synchronous model?

### Describe the solution you'd like

Make everything that can emit impulses - create async tasks, so the difference between sync and async can be hidden. Since as far as people are concerned, it will work pretty much the same way

### Describe alternatives you've considered

Just starting async tasks on everything that gives sync impulse, so I don't have to care about the difference as I create the logic in protoflux

**Comment by shadowpanther** [comment]
The thing is, async that Resonite has is not async in the sense some programming languages have it.

- All async execution is still processed on the same update (*) that the impulse is started, same as sync
- Delays still skip the requested numbed of updates (frames), if you delay by 1 update, the async impulse would continue on the next frame
- The difference between sync and async is that when an async context is created, all variable values and evaluation results are copied into that context, and you can have context-local variables
- Async is not multithreaded, all sync and async impulses are still executed in sequence during the update

(*) Except when the async impulse is delayed, but that's also not exactly true. It looks like the original impulse terminates and adds another impulse to the queue while passing the async context along.

And because the way the async execution treats the context and all inputs, you can't convert everything sync to async easily. Probably that's the reason we still have sync - a lot of legacy items would break if all execution was made async.

**Comment by shadowpanther** [comment]
Most impulse-triggered nodes like Write work both in sync and async. It's the context (the state of inputs) that is processed differently between sync and async.

Resonite's async is not regular programming async. Delays are explicitly processed on the update they should be and not postponed if something is busy. The update (frame) won't end before all async tasks scheduled for it finish or yield via Delay.

And you might also need to use Start Async Task inside the async context to split into another async context tracked separately from the previous one.

**Comment by Nytra** [comment]
Yes as shadowpanther said sometimes you need to explicitly start a new async task if you want to do something with recursion for example. However i can see what you mean otherwise. It would be nice if protoflux could implicitly start async tasks if it detects that it needs to instead of always requiring the node.
